radio/src/audio_arm.h

```diff
@@ -314,38 +317,79 @@ class AudioBufferFifo {
     // returns an empty buffer to be filled wit data and put back into FIFO with audioPushBuffer()
     AudioBuffer * getEmptyBuffer() const
     {
-      return full() ? 0 : &audioBuffers[writeIdx];
+#if defined(PCBSKY9X)
+      AudioBuffer * buffer = &audioBuffers[writeIdx];
+      return buffer->state == AUDIO_BUFFER_FREE ? buffer : NULL;
+#else
+      return full() ? NULL : &audioBuffers[writeIdx];
+#endif
     }
 
     // puts filled buffer into FIFO
     void audioPushBuffer()
     {
-      // AudioBuffer * buffer = &audioBuffers[writeIdx];
       audioDisableIrq();
+#if defined(PCBSKY9X)
+      AudioBuffer * buffer = &audioBuffers[writeIdx];
+      buffer->state = AUDIO_BUFFER_FILLED;
+#endif
       writeIdx = nextBufferIdx(writeIdx);
       bufferFull = (writeIdx == readIdx);
       audioEnableIrq();
-      // buffer->state = AUDIO_BUFFER_FILLED;
     }
 
     // returns a pointer to the audio buffer to be played
-    const AudioBuffer * getNextFilledBuffer() const
+    const AudioBuffer * getNextFilledBuffer()
     {
-      return empty() ? 0 : &audioBuffers[readIdx];
+#if defined(PCBSKY9X)
+      uint8_t idx = readIdx;
+      do {
+        AudioBuffer * buffer = &audioBuffers[idx];
+        if (buffer->state == AUDIO_BUFFER_FILLED) {
+          buffer->state = AUDIO_BUFFER_PLAYING;
+          readIdx = idx;
+          return buffer;
+        }
+        idx = nextBufferIdx(idx);
+      } while (idx != writeIdx);   // this fixes a bug if all buffers are filled
+      return NULL;
+#else
+      return empty() ? NULL : &audioBuffers[readIdx];
+#endif
     }
 
     // frees the last played buffer
     void freeNextFilledBuffer()
     {
       audioDisableIrq();
+#if defined(PCBSKY9X)
+      if (audioBuffers[readIdx].state == AUDIO_BUFFER_PLAYING) {
+        audioBuffers[readIdx].state = AUDIO_BUFFER_FREE;
+        readIdx = nextBufferIdx(readIdx);
+        bufferFull = false;
+      }
+#else
```

Comment:
I only ported the code from 'master' as I knew it works. I will try now to remove this code which gives me headaches